File: src/math/line.ts
import { compare } from "./number";
import Vector2 from "./vector2";

/**
 * Line equation: ax + by + c = 0
 */
class Line {
  public readonly a: number;
  public readonly b: number;
  public readonly c: number;

  public static fromPoints(a: Vector2, b: Vector2): Line {
    return new Line(b.y - a.y, a.x - b.x, b.x * a.y - a.x * b.y);
  }

  public constructor(a: number, b: number, c: number) {
    this.a = a;
    this.b = b;
    this.c = c;
  }

  public getIntersection(other: Line): Vector2 | undefined {
    const coefficientProductDifference = this.a * other.b - other.a * this.b;

    if (compare(coefficientProductDifference, 0) === 0) return undefined;

    return new Vector2(
      (this.b * other.c - other.b * this.c) / coefficientProductDifference,
      (other.a * this.c - this.a * other.c) / coefficientProductDifference
    );
  }
}

export default Line;

File: logic/src/math/line.rs
use super::{compare, Vector2};

/// Line equation: ax + by + c = 0.
#[derive(Debug, Clone)]
pub struct Line {
    pub a: f32,
    pub b: f32,
    pub c: f32,
}

impl Line {
    pub fn from_points(a: Vector2<f32>, b: Vector2<f32>) -> Line {
        if a.equals(&b) {
            panic!("Different points expected");
        }

        Self {
            a: b.y - a.y,
            b: a.x - b.x,
            c: b.x * a.y - a.x * b.y,
        }
    }

    pub fn intersection(&self, other: &Line) -> Option<Vector2<f32>> {
        let coefficient_product_difference = self.a * other.b - other.a * self.b;

        if compare(coefficient_product_difference, 0.0).is_eq() {
            None
        } else {
            Some(Vector2::new(
                (self.b * other.c - other.b * self.c) / coefficient_product_difference,
                (other.a * self.c - self.a * other.c) / coefficient_product_difference,
            ))
        }
    }
}
